BREAKING: Standardize GCP Secret Manager PushSecret metadata format and add CMEK support

janlauber opened the pull request:
# BREAKING: Standardize GCP Secret Manager PushSecret metadata format and add CMEK support

## Problem Statement
1. The GCP Secret Manager provider used a non-standard metadata format for PushSecrets
2. Customer-Managed Encryption Keys (CMEK) support was needed for enhanced security compliance

## Breaking Changes
This PR standardizes the metadata format for PushSecrets to align with other providers. This is a breaking change that requires a minor version bump to 0.12.0.

### Migration Required
Users will need to update their PushSecret specifications to use the new structured metadata format. See the migration guide in the documentation for details.

Old format:
```yaml
metadata:
  annotations: {...}
  labels: {...}
  topics: [...]
```

New format:
```yaml
metadata:
  apiVersion: kubernetes.external-secrets.io/v1alpha1
  kind: PushSecretMetadata
  spec:
    annotations: {...}
    labels: {...}
    topics: [...]
    cmekKeyName: "..." # Optional: for CMEK
```

## Related Issue
Fixes #4209

## Proposed Changes
1. Refactor PushSecret metadata to use standard structured format
2. Add CMEK support through metadata specification
3. Add migration documentation
4. Update tests to use structured metadata format

## Example Usage
```yaml
apiVersion: external-secrets.io/v1alpha1
kind: PushSecret
metadata:
  name: pushsecret-example
spec:
  data:
    - match:
        secretKey: mysecret
        remoteRef:
          remoteKey: my-secret
      metadata:
        apiVersion: kubernetes.external-secrets.io/v1alpha1
        kind: PushSecretMetadata
        spec:
          cmekKeyName: projects/my-project/locations/us-east1/keyRings/my-keyring/cryptoKeys/my-key
          labels:
            managed-by: external-secrets
```

Note: When using CMEK, you must specify a location in the SecretStore as customer-managed encryption keys are region-specific.

## Checklist
- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] Added migration documentation
- [ ] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

Skarlso commented:
@janlauber Hi! Thanks very much!! As a start, please take care of the sonarcloud issues here: https://sonarcloud.io/project/issues?id=external-secrets_external-secrets&pullRequest=4210&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true :)

gusfcarvalho commented:
Regarding implementation - isn’t CMEKs mapped to each individual gcp Sm Secret?

if so, this might be better implemented with PushSecret Metadata, as it allow the same SecretStore to use multiple CMEKs one per secret

Skarlso commented:
I was about to check that out. :D Thanks Gustavo! :)

janlauber commented:
@gusfcarvalho 

Thanks for the feedback! You're absolutely right - I've refactored the implementation to use PushSecret metadata instead of SecretStore configuration. This allows for more flexibility as users can specify different CMEKs for different secrets while using the same SecretStore.

Changes made:
1. Removed `CMEKKeyName` field from SecretStore
2. Added CMEK configuration through PushSecret metadata
3. Updated tests and documentation accordingly

The implementation now looks like this:

```yaml
apiVersion: external-secrets.io/v1alpha1
kind: PushSecret
metadata:
  name: pushsecret-example
spec:
  # ... other fields ...
  data:
    - match:
        secretKey: mysecret
        remoteRef:
          remoteKey: my-secret
      metadata:
        cmekKeyName: projects/my-project/locations/us-east1/keyRings/my-keyring/cryptoKeys/my-key
```

This approach better aligns with GCP's model where encryption is configured per secret. Please let me know if you'd like me to make any additional changes!

Skarlso commented:
Hello!

Nice, nice! Take a look here please https://github.com/external-secrets/external-secrets/pull/4173/ to see how we deal with Metadata. :) We use versioned metadata fields in order to be able to roll out breaking changes or updates to those fields. :)

Skarlso commented:
Honestly, we did a good job with the ADR for the metadata but we seriously need some dev docs on how to use the new metadata field. :)

janlauber commented:
Hi @Skarlso 
I will take a look at it an try to refactor my changes :)

Skarlso commented:
@janlauber Hopefully it's not difficult. You just need to define a struct and use a different utils function. That's all basically. Everything else is provided to you. :)

janlauber commented:
Hi @Skarlso 
I've tried to implement it with the new Metadata and it was quite smooth :)
I've tested it in our GCP environment and it worked!

Looking forward for your feedback

Skarlso commented on Makefile:
```diff
@@ -238,8 +238,8 @@ docker.tag:  ## Emit IMAGE_TAG
 	@$(INFO) docker build
-	echo docker build -f $(DOCKERFILE) . $(DOCKER_BUILD_ARGS) -t $(IMAGE_NAME):$(IMAGE_TAG)
-	DOCKER_BUILDKIT=1 docker build -f $(DOCKERFILE) . $(DOCKER_BUILD_ARGS) -t $(IMAGE_NAME):$(IMAGE_TAG)
+	echo docker build $(if $(PLATFORM),--platform $(PLATFORM),) -f $(DOCKERFILE) . $(DOCKER_BUILD_ARGS) -t $(IMAGE_NAME):$(IMAGE_TAG)
```
Why this change here?

janlauber replied:
I've added the platform args directly but I will remove it since there are docker_build_args which should already handle this.

Skarlso commented on apis/externalsecrets/v1beta1/secretstore_gcpsm_types.go:
```diff
@@ -40,13 +40,14 @@ type GCPWorkloadIdentity struct {
-	Auth GCPSMAuth `json:"auth,omitempty"`
-
 	// ProjectID project where secret is located
-	ProjectID string `json:"projectID,omitempty"`
```
This would be a breaking change for any existing environments that don't set this. I can see that in certain cases we error without a project id as is, but I'm not a 100% sure it's a required value in all situations.

janlauber replied:
I will revert it and try to implement an autodetection of the project provided by the credentials. If I fail I will just leave it as it is.

janlauber replied:
Currently in the official release when you try to add a secretstore or clustersecretstore without a projectID the status of this resource will report an "InvalidProviderConfig" due to the following line:

https://github.com/external-secrets/external-secrets/blob/325cc52790aa10312c6a8ed701d9e2f7d7cabee9/pkg/provider/gcp/secretmanager/provider.go#L149

So I will revert my changes as it works as intended already

Skarlso commented on pkg/provider/gcp/secretmanager/client.go:
```diff
@@ -183,13 +184,35 @@ func (c *Client) PushSecret(ctx context.Context, secret *corev1.Secret, pushSecr
+				Location: c.store.Location,
+			}
+
+			var meta *metadata.PushSecretMetadata[PushSecretMetadataSpec]
+			if pushSecretData.GetMetadata() != nil {
+				var err error
+				meta, err = metadata.ParseMetadataParameters[PushSecretMetadataSpec](pushSecretData.GetMetadata())
+				if err != nil {
+					return fmt.Errorf("failed to parse PushSecret metadata: %w", err)
+				}
+			}
+
+			var cmekKeyName string
+			if meta != nil {
+				cmekKeyName = meta.Spec.CMEKKeyName
+			}
+
+			if cmekKeyName != "" {
+				replica.CustomerManagedEncryption = &secretmanagerpb.CustomerManagedEncryption{
+					KmsKeyName: cmekKeyName,
+				}
+			}
```
```suggestion
			if pushSecretData.GetMetadata() != nil {
				var err error
				meta, err := metadata.ParseMetadataParameters[PushSecretMetadataSpec](pushSecretData.GetMetadata())
				if err != nil {
					return fmt.Errorf("failed to parse PushSecret metadata: %w", err)
				}
				if meta != nil && meta.Spec.CMEKKeyName != "" {
					replica.CustomerManagedEncryption = &secretmanagerpb.CustomerManagedEncryption{
						KmsKeyName: cmekKeyName,
					}
				}
			}
```

In fact, I doubt that once metadata is defined you would end up with an empty CMEKKeyName field, right? But it's fine to check anyways.

Skarlso commented on pkg/provider/gcp/secretmanager/push_secret.go:
```diff
@@ -34,11 +34,12 @@ const (
-type Metadata struct {
-	Annotations map[string]string             `json:"annotations"`
-	Labels      map[string]string             `json:"labels"`
+type PushSecretMetadataSpec struct {
```
Oh snap, I'm so sorry, I didn't realise that metadata here was already a thing.

Thank you for the refactor though. This just means that we'll have to release this in a minor version `0.12.0` because it's a breaking change. Existing specs will need to redefine their metadata field.

janlauber replied:
Yes, you're right! I discovered the existing metadata pattern when implementing CMEK support. This makes the code more consistent with the rest of the codebase and follows the established patterns.

I agree this should be a minor version bump to 0.12.0 since users will need to update their PushSecret specs to use the structured metadata format. I'll update the PR description to:
1. Highlight this as a breaking change
2. Provide clear migration examples for users
3. Note the 0.12.0 version requirement

I will add a section in the docs about migrating from the old metadata format to the new structured format.

Skarlso reviewed (commented):
Thank you for the refactor. Sorry, I didn't realise that metadata was already a thing in GCP. But it's good that we did this, because eventually I would like all providers to move away from arbitrary metadata structure.

Skarlso commented on go.mod:
```diff
@@ -43,7 +43,7 @@ require (
 	github.com/youmark/pkcs8 v0.0.0-20240726163527-a2c0da244d78
 	go.uber.org/zap v1.27.0
-	golang.org/x/crypto v0.30.0
+	golang.org/x/crypto v0.31.0
```
Huh, what prompted this change? 🤔

janlauber replied:
The trivy scan: https://github.com/external-secrets/external-secrets/actions/runs/12329516819/job/34416450574#step:12:217

Skarlso replied:
Oh okay. Neat.
